packages/firestore/src/model/transform_operation.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FieldValue, ServerTimestampValue, ArrayValue } from './field_value';
+import { Timestamp } from '../api/timestamp';
+import * as misc from '../util/misc';
+import { assert } from '../util/assert';
+
+/** Represents a transform within a TransformMutation. */
+export interface TransformOperation {
+  /** Transforms the provided `previousValue`. */
+  transform(previousValue: FieldValue, localWriteTime?: Timestamp): FieldValue;
```

Comment:
Thanks for this refactor!

I played with this a little to see if we could get rid of `localWriteTime` as an argument. I am not entirely happy with it, but here is the result of the preliminary investigation: https://gist.github.com/schmidt-sebastian/6dd9cb25015d14bc96d742c3d27d6fc6

It's a lot of churn, which simplifies some parts of our Server Timestamp handling, but it also adds complexity to other parts (mainly to the serializer).

Reply:
While I would like to like your change, the reason localWriteTime is stored in the MutationBatch rather than in the Mutation or ServerTimestampTransform is so that when you do a batch write, all of your server timestamp sentinels resolve to the same timestamp (i.e. when you committed the batch rather than each one resolving to when you added the mutation to the batch which could theoretically be over a longer time period).

If I didn't miss anything, your change ends up storing a timestamp in both the ServerTimestampTransform and in the MutationBatch, and uses the one in the ServerTimestampTransform for local-application...  but after serializing / deserializing, I think the version in the mutation will end up being replaced with the version in the MutationBatch (so after restarting your app, your local server timestamps may change which isn't good).

So I'm not sure this works out.  We could perhaps make it work by creating the ServerTimestampTransforms during parsing, and then going back and assigning a localWriteTime to each one at the time the MutationBatch is constructed.  That way they're guaranteed to be the same and to match the timestamp in the MutationBatch.

But I'm not sufficiently convinced this is a meaningful improvement. As you ran into with the serializer, our current model translates well to the backend protos / implementation, which has some advantages.  So I'm inclined to stick with the warts we have rather than exchange for new ones.

Reply:
I am glad that I played around with this myself before blindly recommending it. Making the localWriteTime a constructor argument seemed much nicer before I changed the interactions between the mutation batches and the serializer. Thanks for looking at this as well, but I as it stands, I am 100% happy with where this PR landed.